File: sandbox/python/calculator.py
"""Test library for OpenDocs Python extractor."""

from typing import Optional, Union, Literal
from enum import IntEnum


class Calculator:
    """A simple calculator class for testing."""

    PI: float = 3.14159
    """Mathematical constant PI"""

    def __init__(self, precision: int = 2):
        """
        Initialize the calculator.

        Args:
            precision: Number of decimal places for results
        """
        self.precision = precision

    def add(self, a: float, b: float) -> float:
        """
        Add two numbers together.

        Args:
            a: First number
            b: Second number

        Returns:
            The sum of a and b
        """
        return round(a + b, self.precision)

    def subtract(self, a: float, b: float) -> float:
        """
        Subtract b from a.

        Args:
            a: Number to subtract from
            b: Number to subtract

        Returns:
            The difference
        """
        return round(a - b, self.precision)

    @staticmethod
    def multiply(a: float, b: float) -> float:
        """
        Multiply two numbers.

        Args:
            a: First number
            b: Second number

        Returns:
            The product
        """
        return a * b


class Config:
    """Configuration options for the library."""

    def __init__(
        self,
        debug: bool,
        timeout: int,
        max_retries: Optional[int] = None
    ):
        """
        Initialize configuration.

        Args:
            debug: Enable debug mode
            timeout: Timeout in milliseconds
            max_retries: Maximum number of retries (optional)
        """
        self.debug = debug
        self.timeout = timeout
        self.max_retries = max_retries


class LogLevel(IntEnum):
    """Log levels enumeration."""

    DEBUG = 0
    """Debug level logging"""

    INFO = 1
    """Info level logging"""

    WARN = 2
    """Warning level logging"""

    ERROR = 3
    """Error level logging"""


# Type alias for Result
Result = Union[
    dict[Literal["success"], Literal[True]] | dict[Literal["data"], object],
    dict[Literal["success"], Literal[False]] | dict[Literal["error"], str]
]


def greet(name: str) -> str:
    """
    Greet a user by name.

    Args:
        name: Name of the person to greet

    Returns:
        A greeting message
    """
    return f"Hello, {name}!"


def format_currency(amount: float, currency: str = "USD") -> str:
    """
    Format a number as currency.

    Args:
        amount: The amount to format
        currency: Currency code (default: USD)

    Returns:
        Formatted currency string
    """
    return f"{currency} {amount:.2f}"

File: sandbox/go/calculator.go
package testlibrary

import "math"

// Calculator provides basic arithmetic operations
type Calculator struct {
	// Precision determines the decimal precision for operations
	Precision int
}

// Add returns the sum of two numbers
func (c *Calculator) Add(a, b float64) float64 {
	return a + b
}

// Subtract returns the difference between two numbers
func (c *Calculator) Subtract(a, b float64) float64 {
	return a - b
}

// Multiply returns the product of two numbers
func (c *Calculator) Multiply(a, b float64) float64 {
	return a * b
}

// PI represents the mathematical constant pi
const PI = math.Pi

// Config holds configuration options for the library
type Config struct {
	// Debug enables debug mode
	Debug bool `json:"debug"`

	// MaxRetries specifies the maximum number of retries
	MaxRetries int `json:"max_retries,omitempty"`

	// Timeout specifies the timeout in milliseconds
	Timeout int `json:"timeout"`
}

// LogLevel represents logging levels
type LogLevel int

const (
	// DEBUG level for debug messages
	DEBUG LogLevel = iota
	// INFO level for informational messages
	INFO
	// WARN level for warning messages
	WARN
	// ERROR level for error messages
	ERROR
)

// Result represents the result of an operation
type Result struct {
	// Success indicates if the operation was successful
	Success bool `json:"success"`

	// Data holds the result data
	Data interface{} `json:"data,omitempty"`

	// Error holds the error message if any
	Error string `json:"error,omitempty"`
}

// Greet returns a greeting message for the given name
func Greet(name string) string {
	return "Hello, " + name + "!"
}

// FormatCurrency formats an amount as currency
func FormatCurrency(amount float64, currency string) string {
	if currency == "" {
		currency = "USD"
	}
	return currency + " " + math.Round(amount*100)/100
}
